compiler/runtime.go

```diff
@@ -25,6 +26,22 @@ type Evaluator interface {
 	Eval(ctx context.Context, scope Scope) (values.Value, error)
 }
 
+type key int
+
+const runtimeDependenciesKey key = iota
+
+type RuntimeDependencies struct {
```

Comment:
I'd agree and I tried that, but that would force me to change every call into the evaluator to take that extra parameter (or rather, a new `EvaluatorState` struct which would make it easier to add new parameters/state later on). However even the interpreter ends up calling into the evaluator through some builtins so there are just far to much code that would be affected by such a change as part of this PR.

```
type EvaluatorState struct {
    ctx context.Context
   scope Scope
   mem memory.Allocator
}
```

Reply:
I figured you could have the compiler initialize it as an attribute. So `binaryVectorEvaluator` would be like:

```
type binaryVectorEvaluator struct {
    // other attributes
    mem memory.Allocator
}
```

It would then use that during evaluation. The compiler would initialize the evaluator with the memory allocator.

Reply:
I tried that, but it doesn't really change anything as then the evaluator compiler needs access to the allocator and the only way to provide that is by storing it in the `Context`.

Reply:
Yes, and I have tried that but then I ran into callsites like this https://github.com/influxdata/flux/blob/01dcde3e9bae6b6589c3e414333e75b411969875/stdlib/experimental/array/array.go#L152 where the only way to pass in the allocator is via the context again https://github.com/influxdata/flux/blob/01dcde3e9bae6b6589c3e414333e75b411969875/stdlib/experimental/array/array.go#L124 . 

So it doesn't fundamentally change anything, we still need to store the allocator in the context. I do agree we should refactor this such that we can pass the allocator (and any other state we might need) in a type safe way, however it would force changes throughout almost the entire go project since most packages needs to be altered to pass a bespoke struct instead of `Context` (though that struct would have a field holding the `Context`).

If we do want that refactor I can do that before or after this PR depending on what seems best.

Reply:
Ok, I didn't realize that some functions needed access. Let me take a quick look again and then I can hit approve.

Reply:
I took a look and if we're going to inject this into the context this isn't the proper location.

First, the dependency interface isn't really meant for this. It's much more straightforward to use the context directly and not with a struct and a name like `RuntimeDependencies`.

Second, we'll need to add this to the context in `lang.Program` probably. We need to make sure that the allocator is available in the context for both the interpreter and the compiler. The compiler generally gets run inside of the execution context inside of the executor, but the interpreter runs outside of that. Fortunately for us, they're close enough that they seem to use the same allocator.

https://github.com/influxdata/flux/blob/4996f846b917311495e1c83a9ceb4d69d1b16225/lang/compiler.go#L474

It seems to me like we should add the logic inside of the `memory` package instead so that the interpreter, the compiler, and any functions that use it have easy access.

Generally, the Go standard for using context is like this:

```go
type key int

const allocatorKey key = iota

func WithAllocator(ctx context.Context, mem Allocator) context.Context {
    return context.WithValue(ctx, allocatorKey, mem)
}

func GetAllocator(ctx context.Context) Allocator {
    mem, ok := ctx.Value(allocatorKey).(Allocator)
    if !ok {
        return DefaultAllocator
    }
    return mem
}
```

Reply:
Done. I was sure I had added the allocator in `Start` but apparently I hadn't or I removed it for some reason...